### ISSUE TITLE
Error when running recursive forecasting calibration on test set

### ISSUE THREAD

**Issue by AristosK** [business-science/modeltime#185]
Hi, I am running following part of code  based on the 

# * Data Extended ----

FORECAST_HORIZON <- 28

data_extended_tbl <- data_tbl %>%
  
  
  # Global Features / Transformations / Joins
  mutate(SmsTraffic = log1p(SmsTraffic))%>%
  
  # Extend
  group_by(Combo_Id) %>%
  future_frame(SendDate, .length_out = FORECAST_HORIZON, .bind_data = TRUE) %>%
  
  tk_augment_lags(SmsTraffic, .lags = FORECAST_HORIZON, .names = "long_lag") %>%
  tk_augment_slidify(
    .value   = long_lag,
    .f       = ~mean(.x, na.rm=TRUE),
    .period  = c(0.5*FORECAST_HORIZON, FORECAST_HORIZON, FORECAST_HORIZON*2),
    .align   = "center",
    .partial = TRUE
  ) %>%
  
  ungroup()

data_extended_tbl

# * Transformer Function ----

transformer_function <- function(data) {
  data %>%
    group_by(Combo_Id) %>%
    tk_augment_lags(SmsTraffic, .lags = 1:FORECAST_HORIZON) %>%
    ungroup()
}

data_lagged_tbl <- data_extended_tbl %>% transformer_function()

data_prepared_tbl <- data_lagged_tbl %>% drop_na()

data_future_tbl <- data_lagged_tbl %>% filter(is.na(interactions))

# 2.0 TIME SPLIT ----

# * Resamples ----

resamples <- time_series_cv(
  data_prepared_tbl,
  cumulative  = TRUE,
  assess      = FORECAST_HORIZON,
  slice_limit = 1
)

train_tbl <- training(resamples$splits[[1]]) %>%
  group_by(Combo_Id) %>%
  mutate(SmsTraffic = ts_clean_vec(SmsTraffic, period = 7)) %>%
  ungroup()

test_tbl  <- testing(resamples$splits[[1]])

# 2.0 RECIPES ----

# ** Lag Recipe ----

recipe_spec_lag <- recipe(SmsTraffic ~ ., data = train_tbl) %>%
  step_dummy(all_nominal()) %>%
  step_rm(SendDate) %>%
  step_zv(all_predictors())

recipe_spec_lag %>% prep() %>% juice() %>% glimpse()

# ** Calendar + Long-Term Rolling Recipe ----

recipe_spec_calendar <- recipe(SmsTraffic ~ .,data = train_tbl %>% select(-contains("lag"), contains("roll")))%>%
  step_timeseries_signature(SendDate) %>%
  step_normalize(SendDate_index.num, starts_with("SendDate_year"))%>%
  step_dummy(all_nominal(), one_hot = TRUE) %>%
  step_normalize(SendDate_index.num) %>%
  step_rm(SendDate) %>%
  step_zv(all_predictors())

recipe_spec_calendar %>% prep() %>% juice() %>% glimpse()

# ** Hybrid Short-Term Lag + Calendar Recipe ----

recipe_spec_hybrid <- recipe(SmsTraffic ~ ., data = train_tbl) %>%
  step_timeseries_signature(SendDate) %>%
  step_dummy(all_nominal(), one_hot = TRUE) %>%
  step_normalize(SendDate_index.num, starts_with("SendDate_year")) %>%
  step_rm(SendDate) %>%
  step_zv(all_predictors())

recipe_spec_hybrid %>% prep() %>% juice() %>% glimpse()

# A. XGBOOST ----

# * XGBoost Model ----
model_spec_xgboost <- boost_tree(
  mode       = "regression",
  learn_rate = 0.75,
  min_n      = 1,
  tree_depth = 12,
  loss_reduction = 0.001
) %>%
  set_engine("xgboost")

# * Workflows ----

# Lag

wflw_fit_xgboost_lag <- workflow() %>%
  add_model(model_spec_xgboost) %>%
  add_recipe(recipe_spec_lag) %>%
  fit(train_tbl) %>%
  recursive(
    id         = "Combo_Id",
    transform  = transformer_function,
    train_tail = panel_tail(train_tbl, Combo_Id, FORECAST_HORIZON)
  )

# Calendar

wflw_fit_xgboost_calendar <- workflow() %>%
  add_model(model_spec_xgboost) %>%
  add_recipe(recipe_spec_calendar) %>%
  fit(train_tbl)

# Hybrid

wflw_fit_xgboost_hybrid <- workflow() %>%
  add_model(model_spec_xgboost) %>%
  add_recipe(recipe_spec_hybrid) %>%
  fit(train_tbl) %>%
  recursive(
    id = "Combo_Id",
    transform = transformer_function,
    train_tail = panel_tail(train_tbl, Combo_Id, FORECAST_HORIZON)
  )

# * Calibration Results ----
calibration_xgboost_tbl <- modeltime_table(
  wflw_fit_xgboost_lag,
  wflw_fit_xgboost_calendar,
  wflw_fit_xgboost_hybrid
) %>%
  modeltime_calibrate(test_tbl)

calibration_xgboost_tbl %>% modeltime_accuracy(quiet = FALSE)

**The error that appears is the following:**

calibration_xgboost_tbl %>% modeltime_accuracy(quiet = FALSE)
Error: no applicable method for 'group_vars' applied to an object of class "logical"
Error: no applicable method for 'group_vars' applied to an object of class "logical"
ℹ We have detected a possible intermittent series, you can change the default metric set to the extended_forecast_accuracy_metric_set() containing the MAAPE metric, which is more appropriate for this type of series.
# A tibble: 3 × 9
  .model_id .model_desc       .type   mae  mape   mase smape  rmse    rsq
      <int> <chr>             <chr> <dbl> <dbl>  <dbl> <dbl> <dbl>  <dbl>
1         1 RECURSIVE XGBOOST NA    NA       NA NA      NA   NA    NA    
2         2 XGBOOST           Test   1.04   Inf  0.417  39.7  1.56  0.721
3         3 RECURSIVE XGBOOST NA    NA       NA NA      NA   NA    NA

**Comment by AristosK** [comment]
Update: I ran the same recipes this morning and now I was getting this error:

Error: Assigned data `pred_fun(...)` must be compatible with row subscript `new_data$rowid.. == i`.
✖ 54 rows must be assigned.
✖ Assigned data has 57 rows.
ℹ Only vectors of size 1 are recycled.
Error: Assigned data `pred_fun(...)` must be compatible with row subscript `new_data$rowid.. == i`.
✖ 54 rows must be assigned.
✖ Assigned data has 57 rows.
ℹ Only vectors of size 1 are recycled.

**Comment by AristosK** [comment]
So thanks to a hint from Vidar from business science group I was able to solve the problem- of the row mismatch. It was caused from the fact that the timeseries obs per id in the test set didnt have the same number of observations, for example I was having 28 (my test set horizon) and some of them had like 22 or 21. A way to check this after the split is to make a simple count in the test set and ensure all of them have same number of observations, after that you can decide if you wanna alter your initial time frames. If thats not the case then when padding is to add a max end date like:

  # Pad the series
  pad_by_time(
    .date_var = date,
    .by = "day",
    .pad_value = 0,
    .start_date = min(df$date),
    .end_date = max(df$date)
  ) 
  
  This will help create equal length test regardless if there will be some with zero
